File: docs-examples/example-java/src/test/java/io/micronaut/nats/docs/serdes/ProductInfoSerDesSpec.java
package io.micronaut.nats.docs.serdes;

import io.micronaut.context.annotation.Property;
import io.micronaut.nats.testcontainers.Nats;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@MicronautTest
@Property(name = "spec.name", value = "ProductInfoSerDesSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ProductInfoSerDesSpec implements TestPropertyProvider {

    @Test
    void testUsingACustomSerDes(ProductClient productClient, ProductListener productListener) {

// tag::producer[]
        productClient.send(new ProductInfo("small", 10L, true));
        productClient.send(new ProductInfo("medium", 20L, true));
        productClient.send(new ProductInfo(null, 30L, false));
// end::producer[]

        await().atMost(60, SECONDS).until(() ->
                productListener.messages.size() == 3 &&
                productListener.messages.stream().anyMatch(pi -> pi.getCount() == 10L) &&
                productListener.messages.stream().anyMatch(pi -> pi.getCount() == 20L) &&
                productListener.messages.stream().anyMatch(pi -> pi.getCount() == 30L)
        );
    }

    @Override
    public Map<String, String> getProperties() {
        return Nats.getProperties();
    }
}

File: docs-examples/example-python/src/test/python/micronaut/nats/docs/serdes/ProductInfoSerDesSpec.py
from time import sleep
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductInfo import ProductInfo
from .ProductListener import ProductListener


@Property(name="spec.name", value="ProductInfoSerDesSpec")
@MicronautTest(environments=["nats"])
class ProductInfoSerDesSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_using_a_custom_serdes(self):
        # tag::producer[]
        self.product_client.send(ProductInfo("small", 10, True))
        self.product_client.send(ProductInfo("medium", 20, True))
        self.product_client.send(ProductInfo(None, 30, False))
        # end::producer[]

        for _ in range(600):
            if len(self.product_listener.messages) == 3:
                break
            sleep(0.1)
        counts = sorted(int(pi.count) for pi in self.product_listener.messages)
        assert counts == [10, 20, 30]
